<a href="https://colab.research.google.com/github/ThanhVanLe0605/PYTHON-CHO-PH-N-T-CH-D-LI-U-/blob/main/Tinh_ch%E1%BB%89nh_v%C3%A0_l%C3%A0m_s%E1%BA%A1ch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [34]:
import pandas as pd

# định dạng hiển thị 2 chữ số thập phân với kiểu float
pd.set_option("display.float_format", lambda x: format(x, ".2f"))

# đọc dữ liệu từ file
df = pd.read_csv("khach_hang.csv", sep = ";", encoding ="utf-8")

df

Ma        Ho-ten  Tuoi Gioi-tinh Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam  5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu  4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam  6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu  3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam  8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
6    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
7    7  Nguyen Van G    34       Nam        ?  Trung hoc  Da ket hon   
8    8    Dang Van H    27       Nam  5200000  Trung hoc    Doc than   
9    9    Pham Thi I    31        Nu  5500000    Dai hoc  Da ket hon   
10  10   Hoang Thi K    26        Nu  4800000   Cao Dang    Doc than   
11  11    Tran Van L   380       Nam  7200000    Dai hoc  Da ket hon   
12  12  Nguyen Thi M    32        Nu  6000000  Trung hoc    Doc than   
13  13    Tran Van N    29       Nam  5600000    Dai hoc    Doc than   
14  14  Nguyen Thi Q    24        Nu  4200000   Cao dang    Doc than   
15  15    Pham Van S    33       Nam  6700000    Dai hoc  Da ket hon   
16  16   Hoang Thi T    28        Nu  4800000   Cao dang    Doc than   
17  17  Nguyen Van U    37       Nam  7000000    Dai hoc  Da ket hon   
18  18      Le Van X    41       Nam  8000000    Dai hoc  Da ket hon   
19  19    Pham Thi Y    27        Nu  5500000   Cao dang    Doc than   
20  20  Nguyen Thi Z    36        Nu  6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4               ?      720  
5   Dang lam viec      710  
6   Dang lam viec      710  
7   Dang lam viec      760  
8   Dang lam viec      680  
9   Dang lam viec      700  
10    That nghiep      640  
11  Dang lam viec      780  
12  Dang lam viec        ?  
13  Dang lam viec      700  
14    That nghiep      610  
15  Dang lam viec      730  
16  Dang lam viec      670  
17  Dang lam viec      750  
18  Dang lam viec        ?  
19  Dang lam viec      700  
20    That nghiep      720

Trước tiên, dùng **info()** để xem thông tin cơ bản của tậ dữ liệu cũng nhưu có cái nhìn tổng quan về tập dữ liệu

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Ma         21 non-null     int64 
 1   Ho-ten     21 non-null     object
 2   Tuoi       21 non-null     int64 
 3   Gioi-tinh  21 non-null     object
 4   Thu-nhap   21 non-null     object
 5   Trinh-do   21 non-null     object
 6   Hon-nhan   21 non-null     object
 7   Viec-lam   21 non-null     object
 8   Tin-dung   21 non-null     object
dtypes: int64(2), object(7)
memory usage: 1.6+ KB


Tập dữ liệu hiện tại đang gặp các vấn đề như sau:
- **Duplicate value (trùng lắp dữ liệu)**: dòng thứ 5 và thứ 6
- **Missing value**
- **Wrong format**: cột "Thu_nhap" và "Tin_dung" có định dạng kiểu không phud hợp (dữ liệu nên kiểu float thay vì object)
- **Wrong data**: Cột "Trinh-do"có 1 số dữ lệu không đồng nhất ("Dai hoc"và "Dai Hoc", "Cao dang"và "CaoDang") . Cột "Tuoi", khách hàng ("Ma- 11) có giá trị không phù hợp (380 tuổi)


# 7.1. Làm sạch dữ liệu

## 7.1.1. Dữ liệu trùng lắp

### a. Xác định dữ liệu trùng

Để kiểm tra giá trị trùng lắp, chúng ta có thể sử dụng phương thứu $duplicated()$ trong thưu viện pandas. Kết quả trả về giá trị "True" nghĩa là dòng dữ liệu đó bị trùng, "False" nghĩa là không bị trùng

In [35]:
df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
dtype: bool

Quan sát kết quả, chúng ta thấy rằng là hàng 6 (có giá trị "True"), tức hàng 6 có giá trị bị trùnglắp và cần được xử lý. Lưu ý rằng, cách xác định này chỉ áp dụng khi dữ liệu có kích thước nhỏ. Khidữ liệu có kích thước lớn, việc quan sát như vậy trở nên không khả thi. Thay vào đó chúng ta có thểđếm số lượng dòng bị trùng bằng cách tính tổng số lượng các dòng có giá trị "True".Đếm xem có bao nhiêu dòng dữ liệu bị trùng:

In [5]:
df.duplicated().sum()

np.int64(1)

Vậy dữ liệu chứa 1 dòng giá trị trùng lắp. Tiếp theo chúng ta sẽ thực hiện xử lý giá trị trùng lắp

### b. Loại bỏ dữ liệu trùng

Các đơn giản nhất để xử lý các giá trị trùng lắp đó là loại bỏ chúng ra khỏi tập dữ liệu. Chúng tasử dụng phương drop_duplicates() trong thư viện Pandas để loại bỏ giá trị trùng. Chú ý sử dụnginplace=True để tác động vào trong dữ liệu.

In [36]:
# loại bỏ giá trị trùng lắp
df.drop_duplicates(inplace = True)

Kiểm tra trùng lắp lần nữa

In [7]:
df.duplicated().sum()

np.int64(0)

Vậy dữ liệu đã không còn trùng lắp. Bằng cách xem lại dữ liệu, chúng ta thấy rằng, dòng dữ liệuthứ 6 (index = 6) (dòng bị trùng lắp trước đó) đã bị xoá.

In [8]:
df

Ma        Ho-ten  Tuoi Gioi-tinh Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam  5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu  4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam  6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu  3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam  8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
7    7  Nguyen Van G    34       Nam        ?  Trung hoc  Da ket hon   
8    8    Dang Van H    27       Nam  5200000  Trung hoc    Doc than   
9    9    Pham Thi I    31        Nu  5500000    Dai hoc  Da ket hon   
10  10   Hoang Thi K    26        Nu  4800000   Cao Dang    Doc than   
11  11    Tran Van L   380       Nam  7200000    Dai hoc  Da ket hon   
12  12  Nguyen Thi M    32        Nu  6000000  Trung hoc    Doc than   
13  13    Tran Van N    29       Nam  5600000    Dai hoc    Doc than   
14  14  Nguyen Thi Q    24        Nu  4200000   Cao dang    Doc than   
15  15    Pham Van S    33       Nam  6700000    Dai hoc  Da ket hon   
16  16   Hoang Thi T    28        Nu  4800000   Cao dang    Doc than   
17  17  Nguyen Van U    37       Nam  7000000    Dai hoc  Da ket hon   
18  18      Le Van X    41       Nam  8000000    Dai hoc  Da ket hon   
19  19    Pham Thi Y    27        Nu  5500000   Cao dang    Doc than   
20  20  Nguyen Thi Z    36        Nu  6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4               ?      720  
5   Dang lam viec      710  
7   Dang lam viec      760  
8   Dang lam viec      680  
9   Dang lam viec      700  
10    That nghiep      640  
11  Dang lam viec      780  
12  Dang lam viec        ?  
13  Dang lam viec      700  
14    That nghiep      610  
15  Dang lam viec      730  
16  Dang lam viec      670  
17  Dang lam viec      750  
18  Dang lam viec        ?  
19  Dang lam viec      700  
20    That nghiep      720

Cuối cùng, chúng ta sử dụng phương thức reset_index() để thiết lặp lại các giá trị của cột indexcho phù hợp

In [37]:
df.reset_index(drop = True, inplace = True)
df

Ma        Ho-ten  Tuoi Gioi-tinh Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam  5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu  4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam  6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu  3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam  8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam        ?  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam  5200000  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu  5500000    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu  4800000   Cao Dang    Doc than   
10  11    Tran Van L   380       Nam  7200000    Dai hoc  Da ket hon   
11  12  Nguyen Thi M    32        Nu  6000000  Trung hoc    Doc than   
12  13    Tran Van N    29       Nam  5600000    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu  4200000   Cao dang    Doc than   
14  15    Pham Van S    33       Nam  6700000    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu  4800000   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam  7000000    Dai hoc  Da ket hon   
17  18      Le Van X    41       Nam  8000000    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu  5500000   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu  6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4               ?      720  
5   Dang lam viec      710  
6   Dang lam viec      760  
7   Dang lam viec      680  
8   Dang lam viec      700  
9     That nghiep      640  
10  Dang lam viec      780  
11  Dang lam viec        ?  
12  Dang lam viec      700  
13    That nghiep      610  
14  Dang lam viec      730  
15  Dang lam viec      670  
16  Dang lam viec      750  
17  Dang lam viec        ?  
18  Dang lam viec      700  
19    That nghiep      720

## 7.1.2. Dữ liệu bị thiếu

### a. Xác định giá trị bị thuế

Trong tập dữ liệu, các ô dữ liệu bị thiếu được hiển thị với dấu "?".

Chúng ta sẽ thay thể các "?"bằng giá trị NaN (Not a Number) - giá trị bị thiếu mặc định củaPython vì lý do tốc độ tính toán và sự thuận tiện. Ở đây chúng ta sử dụng phương thức $replace(A, B,inplace = True)$ để thay thế giá trị A bởi B

In [38]:
import numpy as np
# thay thế "?" bằng NaN
df.replace("?", np.nan, inplace = True)


In [11]:
df

Ma        Ho-ten  Tuoi Gioi-tinh Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam  5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu  4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam  6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu  3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam  8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam      NaN  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam  5200000  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu  5500000    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu  4800000   Cao Dang    Doc than   
10  11    Tran Van L   380       Nam  7200000    Dai hoc  Da ket hon   
11  12  Nguyen Thi M    32        Nu  6000000  Trung hoc    Doc than   
12  13    Tran Van N    29       Nam  5600000    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu  4200000   Cao dang    Doc than   
14  15    Pham Van S    33       Nam  6700000    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu  4800000   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam  7000000    Dai hoc  Da ket hon   
17  18      Le Van X    41       Nam  8000000    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu  5500000   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu  6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4             NaN      720  
5   Dang lam viec      710  
6   Dang lam viec      760  
7   Dang lam viec      680  
8   Dang lam viec      700  
9     That nghiep      640  
10  Dang lam viec      780  
11  Dang lam viec      NaN  
12  Dang lam viec      700  
13    That nghiep      610  
14  Dang lam viec      730  
15  Dang lam viec      670  
16  Dang lam viec      750  
17  Dang lam viec      NaN  
18  Dang lam viec      700  
19    That nghiep      720

Sau khi các giá trị bị thiếu được đưa về dạng mặc định trong python. Chúng ta sử dụng phươngthức $isnull()$ để xác định các giá trị còn thiếu này. Giá trị trả về là một giá trị boolean. Giá trị"True"cónghĩa là giá trị là giá trị bị thiếu trong khi "False"có nghĩa là giá trị không phải là giá trị bị thiếu

In [39]:
missing_data = df.isnull()

missing_data.head(5)

Ma  Ho-ten   Tuoi  Gioi-tinh  Thu-nhap  Trinh-do  Hon-nhan  Viec-lam  \
0  False   False  False      False     False     False     False     False   
1  False   False  False      False     False     False     False     False   
2  False   False  False      False     False     False     False     False   
3  False   False  False      False     False     False     False     False   
4  False   False  False      False     False     False     False      True   

   Tin-dung  
0     False  
1     False  
2     False  
3     False  
4     False

Đếm các giá trị còn thiếu trong mỗi cộtChúng ta có thể nhanh chóng tìm ra số lượng giá trị còn thiếu trong mỗi cột bằng cách sử dụng vònglặp for trong Python. Như đã đề cập ở trên, "True"đại diện cho một giá trị bị thiếu và "False"có nghĩalà giá trị có trong tập dữ liệu. Do đó, trong phần thân của vòng lặp for, chúng ta sẽ sử dụng phươngthức .value_counts() để đếm số lượng giá trị "True".

In [40]:
for column in missing_data.columns.values:
  print(missing_data[column].value_counts())

Ma
False    20
Name: count, dtype: int64
Ho-ten
False    20
Name: count, dtype: int64
Tuoi
False    20
Name: count, dtype: int64
Gioi-tinh
False    20
Name: count, dtype: int64
Thu-nhap
False    19
True      1
Name: count, dtype: int64
Trinh-do
False    20
Name: count, dtype: int64
Hon-nhan
False    20
Name: count, dtype: int64
Viec-lam
False    19
True      1
Name: count, dtype: int64
Tin-dung
False    18
True      2
Name: count, dtype: int64


Chúng ta đạt được kết quả:•"Thu-nhap": 1 ô dữ liệu thiếu.•"Viec-lam": 1 ô dữ liệu thiếu.•"Tin-dung": 2 ô dữ liệu thiếu

### b. Xử lý giá trị thiếu

Các phương pháp xử lý giá trị bị thiếu:
•Xoá dữ liệu
–Xoá cả dòng
–Xoá cả cột

•Thay thế dữ liệu
–Thay thế bởi giá trị trung bình (mean).
–Thay thế bởi giá trị trung vị (median).
–Thay thế bởi giá trị xuất hiện nhiều nhất (mode).

Lưu ý: Chúng ta chỉ thực hiện phương pháp xoá cả cột khi mà hầu hết các ô giá trị trong cột đềubị trống. Trong tập dữ liệu hiện tại của chúng ta, không có cột nào mà có các giá trị hầu hết là trống.

Ở đây, chúng ta minh hoạt áp dụng phương pháp xoá cho cột "Tin-dung", phương pháp thay thếdùng giá trị trung bình cho cột "Thu-nhap"(numeric feature) và thay thế dùng giá trị xuất hiện nhiềunhất cho cột "Viec-lam"(categorical feature)

**Xóa dữ liệu**

Trong trường hợp, nếu mục tiêu của bài toán hiện tại là xem xét các yếu tố ảnh hưởng đến điểm tíndụng của khách hàng hoặc trong trường hợp cần dự đoán điểm tín dụng của một khách hàng, thì giátrị của cột "Tin-dung"được xem là biến mục tiêu (target variable), đóng vai trò quan trọng. Do đó,khi giá trị của cột "Tin-dung"bị thiếu, chúng ta nên xoá những đó đi. (Nếu sử dụng phương phápthay thế giá trị (bằng mean, median hay mode), thì khả năng cao nó sẽ làm giảm hiệu quả các mô hình dự đoán)

In [14]:
## xóa dòng dữ liệu mà giá trị tại cột Tin_dung là NaN
df.dropna(subset=["Tin-dung"], axis =0, inplace = True)


In [15]:
df

Ma        Ho-ten  Tuoi Gioi-tinh Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam  5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu  4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam  6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu  3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam  8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu  5500000    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam      NaN  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam  5200000  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu  5500000    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu  4800000   Cao Dang    Doc than   
10  11    Tran Van L   380       Nam  7200000    Dai hoc  Da ket hon   
12  13    Tran Van N    29       Nam  5600000    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu  4200000   Cao dang    Doc than   
14  15    Pham Van S    33       Nam  6700000    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu  4800000   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam  7000000    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu  5500000   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu  6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4             NaN      720  
5   Dang lam viec      710  
6   Dang lam viec      760  
7   Dang lam viec      680  
8   Dang lam viec      700  
9     That nghiep      640  
10  Dang lam viec      780  
12  Dang lam viec      700  
13    That nghiep      610  
14  Dang lam viec      730  
15  Dang lam viec      670  
16  Dang lam viec      750  
18  Dang lam viec      700  
19    That nghiep      720

**Thay thế dữ liệu**

**Thay thế bằng trung bình**

Lưu ý, hiện tại các giá trị trong cột "Thu-nhap"vẫn đang là kiểu "object". Do đó, chưa thể tính giá trịtrung bình trực tiếp mà cần chuyển kiểu "float"rồi mới tính giá trị trung bình (mean). Sau khi tính được giá trị trung bình, chúng ta thay thế giá trị NaN bằng cách sử dụng phương thức $fillna()$ trong Pandas.

In [41]:
x = df['Thu-nhap'].astype("float").mean()
df["Thu-nhap"].fillna(x, inplace = True)

/tmp/ipython-input-56518716.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Thu-nhap"].fillna(x, inplace = True)


**Thay thế bằng giá trị xuất hiện nhiều nhất**

Trong trường hợp cột "Viec-lam", chúng ta thấy phần lớn các khách hàng đều đang có công việc. Dođó, nếu thay thế giá trị bị thiếu NaN bằng giá trị "Dang lam viec"thì khả năng cao là phù hợp vớithực tế. Cụ thể:

In [42]:
df['Viec-lam'].value_counts()

Viec-lam
Dang lam viec    14
That nghiep       4
Sinh vien         1
Name: count, dtype: int64

Để đảm bảo tính tổng quát, chúng ta sẽ thực hiện thay thế giá trị NaN của cột "Viec-lam"bằngcách sử dụng giá trị xuất hiện nhiều nhất (mode).

In [43]:
x = df["Viec-lam"].mode()[0]
df['Viec-lam'].fillna(x, inplace = True)

/tmp/ipython-input-1228006076.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Viec-lam'].fillna(x, inplace = True)


In [44]:
df


Ma        Ho-ten  Tuoi Gioi-tinh   Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam    5000000    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu    4500000   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam    6000000    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu    3500000   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam    8000000  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu    5500000    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam 5778947.37  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam    5200000  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu    5500000    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu    4800000   Cao Dang    Doc than   
10  11    Tran Van L   380       Nam    7200000    Dai hoc  Da ket hon   
11  12  Nguyen Thi M    32        Nu    6000000  Trung hoc    Doc than   
12  13    Tran Van N    29       Nam    5600000    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu    4200000   Cao dang    Doc than   
14  15    Pham Van S    33       Nam    6700000    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu    4800000   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam    7000000    Dai hoc  Da ket hon   
17  18      Le Van X    41       Nam    8000000    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu    5500000   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu    6800000    Dai hoc  Da ket hon   

         Viec-lam Tin-dung  
0   Dang lam viec      750  
1     That nghiep      680  
2   Dang lam viec      800  
3       Sinh vien      600  
4   Dang lam viec      720  
5   Dang lam viec      710  
6   Dang lam viec      760  
7   Dang lam viec      680  
8   Dang lam viec      700  
9     That nghiep      640  
10  Dang lam viec      780  
11  Dang lam viec      NaN  
12  Dang lam viec      700  
13    That nghiep      610  
14  Dang lam viec      730  
15  Dang lam viec      670  
16  Dang lam viec      750  
17  Dang lam viec      NaN  
18  Dang lam viec      700  
19    That nghiep      720

## 7.1.3. Dữ liệu sai định dạng

### a. Kiểm tra dữ liệu sai định dạng

Bước tiếp theo trong việc làm sạch dữ liệu là kiểm tra và đảm bảo rằng tất cả dữ liệu đều ở đúngđịnh dạng. Việc dữ liệu không phù hợp định dạng có thể gây ra sai sót trong quá trình phân tích.

Trong Pandas, chúng ta sử dụng:
•.dtypeđể kiểm tra kiểu dữ liệu.
•.astype()để thay đổi kiểu dữ liệu.Trước tiên, chúng ta liệt kê để xem các kiểu dữ liệu của từng cột:

In [45]:
df.dtypes

Ma            int64
Ho-ten       object
Tuoi          int64
Gioi-tinh    object
Thu-nhap     object
Trinh-do     object
Hon-nhan     object
Viec-lam     object
Tin-dung     object
dtype: object

Dữ liệu hiện tại tồn tại 2 cột có dữ liệu chưa phù hợp. Cột "Thu-nhap"và "Tin-dung"nên là kiểudữ liệu số thực ("float") thay vì kiểu đối tượng ("object")

### b. Xử lý dữ liệu sai định dạng

Sau đây, chúng ta sẽ thực hiện chuyển đổi dữ liệu bằng cách sử dụng phương thức astype() trong pandas

In [46]:
df['Thu-nhap'] = df['Thu-nhap'].astype('float')
df['Tin-dung'] = df['Tin-dung'].astype('float')

In [22]:
df.dtypes

Ma             int64
Ho-ten        object
Tuoi           int64
Gioi-tinh     object
Thu-nhap     float64
Trinh-do      object
Hon-nhan      object
Viec-lam      object
Tin-dung     float64
dtype: object

Vậy là cột "Thu-nhap"và "Tin-dung"đã được đưa về kiểu dữ liệu phù hợp.

In [47]:
df

Ma        Ho-ten  Tuoi Gioi-tinh   Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam 5000000.00    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu 4500000.00   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam 6000000.00    Dai Hoc  Da ket hon   
3    4    Phan Thi D    22        Nu 3500000.00   Cao Dang    Doc than   
4    5   Hoang Van E    40       Nam 8000000.00  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu 5500000.00    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam 5778947.37  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam 5200000.00  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu 5500000.00    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu 4800000.00   Cao Dang    Doc than   
10  11    Tran Van L   380       Nam 7200000.00    Dai hoc  Da ket hon   
11  12  Nguyen Thi M    32        Nu 6000000.00  Trung hoc    Doc than   
12  13    Tran Van N    29       Nam 5600000.00    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu 4200000.00   Cao dang    Doc than   
14  15    Pham Van S    33       Nam 6700000.00    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu 4800000.00   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam 7000000.00    Dai hoc  Da ket hon   
17  18      Le Van X    41       Nam 8000000.00    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu 5500000.00   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu 6800000.00    Dai hoc  Da ket hon   

         Viec-lam  Tin-dung  
0   Dang lam viec    750.00  
1     That nghiep    680.00  
2   Dang lam viec    800.00  
3       Sinh vien    600.00  
4   Dang lam viec    720.00  
5   Dang lam viec    710.00  
6   Dang lam viec    760.00  
7   Dang lam viec    680.00  
8   Dang lam viec    700.00  
9     That nghiep    640.00  
10  Dang lam viec    780.00  
11  Dang lam viec       NaN  
12  Dang lam viec    700.00  
13    That nghiep    610.00  
14  Dang lam viec    730.00  
15  Dang lam viec    670.00  
16  Dang lam viec    750.00  
17  Dang lam viec       NaN  
18  Dang lam viec    700.00  
19    That nghiep    720.00

## 7.1.4. Dữ liệu bị sai giá trị

Giá trị không đồng nhất

#### a. Kiểm tra giá trị không đồng nhất

In [48]:
cate_features = [
    "Gioi-tinh",
    "Trinh-do",
    "Hon-nhan",
    "Viec-lam"
]

for column in cate_features:
  print(f"{column}: \n{df[column].unique()}")

Gioi-tinh: 
['Nam' 'Nu']
Trinh-do: 
['Dai hoc' 'Cao dang' 'Dai Hoc' 'Cao Dang' 'Trung hoc']
Hon-nhan: 
['Da ket hon' 'Doc than']
Viec-lam: 
['Dang lam viec' 'That nghiep' 'Sinh vien']


#### b.Xử lý giá trị không đồng nhất


Chúng ta thấy rằng, cột "Trinh-do"có các giá trị không đồng nhất với nhau. Điều này có thể dẫn đếnsai sót khi thực hiện một số phân tích như thống kê về số lượng khách hàng theo trình độ chẳng hạn.

Để giải quyết điều này, chúng ta cần chuyển những giá trị về dạng thống nhất. Ở đây, chúng ta sẽthực hiện chuyển tất cả các giá trị "Dai Hoc"thành "Dai hoc"và "Cao Dang"thành "Cao dang"trongcột "Trinh-do"

In [49]:
df['Trinh-do']= df['Trinh-do'].replace({"Dai Hoc": "Dai hoc",
                                       "Cao Dang": "Cao dang"})


Thực hiện kiểm tra lại cột "Trinh-do"

In [50]:
df['Trinh-do'].unique()

array(['Dai hoc', 'Cao dang', 'Trung hoc'], dtype=object)

In [30]:
df

Ma        Ho-ten  Tuoi Gioi-tinh   Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam 5000000.00    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu 4500000.00   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam 6000000.00    Dai hoc  Da ket hon   
3    4    Phan Thi D    22        Nu 3500000.00   Cao dang    Doc than   
4    5   Hoang Van E    40       Nam 8000000.00  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu 5500000.00    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam 5635294.12  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam 5200000.00  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu 5500000.00    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu 4800000.00   Cao dang    Doc than   
10  11    Tran Van L   380       Nam 7200000.00    Dai hoc  Da ket hon   
12  13    Tran Van N    29       Nam 5600000.00    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu 4200000.00   Cao dang    Doc than   
14  15    Pham Van S    33       Nam 6700000.00    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu 4800000.00   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam 7000000.00    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu 5500000.00   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu 6800000.00    Dai hoc  Da ket hon   

         Viec-lam  Tin-dung  
0   Dang lam viec    750.00  
1     That nghiep    680.00  
2   Dang lam viec    800.00  
3       Sinh vien    600.00  
4   Dang lam viec    720.00  
5   Dang lam viec    710.00  
6   Dang lam viec    760.00  
7   Dang lam viec    680.00  
8   Dang lam viec    700.00  
9     That nghiep    640.00  
10  Dang lam viec    780.00  
12  Dang lam viec    700.00  
13    That nghiep    610.00  
14  Dang lam viec    730.00  
15  Dang lam viec    670.00  
16  Dang lam viec    750.00  
18  Dang lam viec    700.00  
19    That nghiep    720.00

Giá trị không phù hợp

#### a. Kiểm tra giá trị không phù hợp

Đôi khi chúng ta có thể phát hiện ra dữ liệu sai bằng cách xem tập dữ liệu và so sánh với mong đợicủa chúng ta về tập dữ liệu đó sẽ như thế nào.Nếu xem lại tập dữ liệu, chúng ta có thể thấy rằng, tuổi là 380 là không phù hợp. Nếu nhìn theokhía cạnh sự mong đợi của chúng ta, độ tuổi phù hợp của một khách hàng chỉ nên từ 18 - 100 tuổi. Ở góc nhìn khác, hầu hết các khách hàng có độ tuổi nằm trong khoảng từ 22-40 tuổi (xem xét phân bốcủa nhóm tuổi). Do đó, giá trị 380 ở cột "Tuổi"có khả năng cao là một giá trị bị sai.

Trong trường hợp này, giá trị 380 trở thành giá trị ngoại lệ (outlier), dẫn đến sự sai lệch trong kếtquả phân tích và giảm hiệu quả của các mô hình được xây dựng nếu như không được xử lý

#### b. Xử lý giá trị không phù hợp

Một cách để xử lý các giá trị bị sai là thay thế chúng bằng một giá trị khác.Trong tập dữ liệu ví dụ hiện tại, ở dòng 10 (tương ứng Khách hàng có "Mã KH- 11), rất có thể đây là lỗi đánh máy và giá trị phải là "38"thay vì "380" và chúng ta chỉ cần thay thế giá trị "380" bởi 38

**Thay thế giá trị**

In [51]:
df.loc[10, "Tuoi"] = 38
df

Ma        Ho-ten  Tuoi Gioi-tinh   Thu-nhap   Trinh-do    Hon-nhan  \
0    1  Nguyen Van A    30       Nam 5000000.00    Dai hoc  Da ket hon   
1    2    Tran Thi B    28        Nu 4500000.00   Cao dang  Da ket hon   
2    3     Le Minh C    35       Nam 6000000.00    Dai hoc  Da ket hon   
3    4    Phan Thi D    22        Nu 3500000.00   Cao dang    Doc than   
4    5   Hoang Van E    40       Nam 8000000.00  Trung hoc  Da ket hon   
5    6  Nguyen Thi F    29        Nu 5500000.00    Dai hoc  Da ket hon   
6    7  Nguyen Van G    34       Nam 5778947.37  Trung hoc  Da ket hon   
7    8    Dang Van H    27       Nam 5200000.00  Trung hoc    Doc than   
8    9    Pham Thi I    31        Nu 5500000.00    Dai hoc  Da ket hon   
9   10   Hoang Thi K    26        Nu 4800000.00   Cao dang    Doc than   
10  11    Tran Van L    38       Nam 7200000.00    Dai hoc  Da ket hon   
11  12  Nguyen Thi M    32        Nu 6000000.00  Trung hoc    Doc than   
12  13    Tran Van N    29       Nam 5600000.00    Dai hoc    Doc than   
13  14  Nguyen Thi Q    24        Nu 4200000.00   Cao dang    Doc than   
14  15    Pham Van S    33       Nam 6700000.00    Dai hoc  Da ket hon   
15  16   Hoang Thi T    28        Nu 4800000.00   Cao dang    Doc than   
16  17  Nguyen Van U    37       Nam 7000000.00    Dai hoc  Da ket hon   
17  18      Le Van X    41       Nam 8000000.00    Dai hoc  Da ket hon   
18  19    Pham Thi Y    27        Nu 5500000.00   Cao dang    Doc than   
19  20  Nguyen Thi Z    36        Nu 6800000.00    Dai hoc  Da ket hon   

         Viec-lam  Tin-dung  
0   Dang lam viec    750.00  
1     That nghiep    680.00  
2   Dang lam viec    800.00  
3       Sinh vien    600.00  
4   Dang lam viec    720.00  
5   Dang lam viec    710.00  
6   Dang lam viec    760.00  
7   Dang lam viec    680.00  
8   Dang lam viec    700.00  
9     That nghiep    640.00  
10  Dang lam viec    780.00  
11  Dang lam viec       NaN  
12  Dang lam viec    700.00  
13    That nghiep    610.00  
14  Dang lam viec    730.00  
15  Dang lam viec    670.00  
16  Dang lam viec    750.00  
17  Dang lam viec       NaN  
18  Dang lam viec    700.00  
19    That nghiep    720.00

Đối với các tập dữ liệu nhỏ, chúng ta có thể thay thế từng dữ liệu sai, nhưng đối với các tập dữ liệu lớn thì điều này thực sự không khả thi

Để thay thế dữ liệu sai cho các tập dữ liệu lớn hơn, chúng ta có thể thiết lập một số quy tắc, vídụ: đặt một số ranh giới cho các giá trị và thay thế bất kỳ giá trị nào nằm ngoài ranh giới

Ví dụ, chúng ta muốn thay thế các giá trị tuổi lớn hơn 100 bằng giá trị 100 và nhỏ hơn 18 bằng giá trị 18




In [52]:
for x in df.index:
  if df.loc[x, "Tuoi"]  > 100:
    df.loc[x, "Tuoi"] = 100
  elif df.loc[x, "Tuoi"] < 18:
    df.loc[x, "Tuoi"] = 18

**Thay thế**

Một cách khác để xử lý dữ liệu sai là loại bỏ các dòng chứa dữ liệu sai. Bằng cách này, chúng takhông cần phải tìm ra giá trị để thay thế chúng.

Ví dụ, chúng ta muốn xoá đi các dòng mà giá trị tuổi lớn hơn 100 hoặc nhỏ hơn 18.

In [ ]:
for x in df.index:
  if df.loc[x, "Tuoi"] < 18 or df.loc[x, "Tuoi"] > 100:
    df.drop(x, inplace = True)

Lưu ý: Phương pháp thay thế và loại bỏ khi xử lý dữ liệu bị sai (wrong data) đều có những ưuđiểm và hạn chế riêng

**Phương pháp thay thế**

Ưu điểm:

- Giữ nguyên kích thước tập dữ liệu: Phương pháp thay thế cho phép giữ nguyênkích thước tập dữ liệu ban đầu, nghĩa là không có mất mát dữ liệu

- Bảo toàn thông tin: Thay vì loại bỏ dữ liệu sai, phương pháp thay thế giữ lạithông tin và có thể cải thiện tính đại diện của tập dữ liệu

Hạn chế
- Khó xác định giá trị thay thế: Đôi khi việc xác định giá trị thay thế phù hợp có thể khá khó, đặc biệt khi dữ liệu sai không rõ ràng

**Phương pháp Loại Bỏ**

Ưu điểm:
- Loại bỏ được nhưng dữ liệu sai: giúp làm sạch tập dữ liệu và đảm bảo tính chính xác của phân tích
- Dễ thực hiện và dễ kiểm tra lại

Hạn chế
- Mất mát thông tin: Loại bỏ dữ liệu có thể dẫn đến mất mát thông tin quý báu và làm giảm độ đại diện của tập dữ liệu.
- Sai lệch (bias): Nếu loại bỏ quá nhiều dữ liệu hoặc dữ liệu quan trọng, có thể gây ra sai lệch trong phân tích và dự đoán

Tóm lại, phương pháp thay thế và loại bỏ đều có ưu điểm và hạn chế riêng. Quyết định sử dụngphương pháp nào phụ thuộc vào ngữ cảnh cụ thể, mục tiêu phân tích, và tính độ chính xác của dữliệu. Trong một số trường hợp, có thể kết hợp cả hai phương pháp để đảm bảo tính chính xác và bảotoàn thông tin quý báu

**Lưu lại**

Cuối cùng, sau giai đoạn làm sạch, chúng ta lưu lại dữ liệu đã làm sạch vào file *.csv như sau:

In [53]:
df.to_csv("clean_khach_hang.csv", index=False)

# 7.2. TINH CHỈNH DỮ LIỆU

Trong quá trình phân tích và xây dựng các mô hình dự đoán, dữ liệu thường được thu thập từ nhiềunguồn khác nhau và có định dạng, đặc điểm riêng biệt. Để làm cho dữ liệu này có thể sử dụng hiệuquả trong các mô hình máy học và thống kê, chúng ta cần chuyển đổi dữ liệu từ các định dạng khácnhau về các định dạng hoặc chuẩn phù hợp cho mục tiêu sử dụng. Thông thương dữ liệu có thể được chia thành 2 nhóm:
- Dữ liệu số (nummeric data)
- Dữ liệu phân loại (categorical data)

## 7.2.1. Chuyển đổi giá trị dữ liệu

Trong tập dữ liệu về Khách hàng đang sử dụng, cột "Thu-nhap"được biểu thị bằng đơn vị VND (ViệtNam Đồng). Giả sử chúng ta đang phát triển một ứng dụng ở một nhóm các quốc gia chấp nhậnđơn vị USD (United States Dollar). Chúng ta cần thực hiện một quá trình chuyển đổi dữ liệu (datatransformation) để chuyển đổi từ VND sang USD.

Giả sử công thức chuyển đổi đơn vị là:

> 1 USD = 24.000 VND



Chuyển đổi giá trị cột thu nhập từ VND sang USD

In [54]:
import pandas as pd
df = pd.read_csv("clean_khach_hang.csv")

df["Thu-nhap"] = df["Thu-nhap"] / 24000

df.rename(columns={"Thu-nhap": "Thu-nhap-USD"}, inplace = True)

df.head()

Ma        Ho-ten  Tuoi Gioi-tinh  Thu-nhap-USD   Trinh-do    Hon-nhan  \
0   1  Nguyen Van A    30       Nam        208.33    Dai hoc  Da ket hon   
1   2    Tran Thi B    28        Nu        187.50   Cao dang  Da ket hon   
2   3     Le Minh C    35       Nam        250.00    Dai hoc  Da ket hon   
3   4    Phan Thi D    22        Nu        145.83   Cao dang    Doc than   
4   5   Hoang Van E    40       Nam        333.33  Trung hoc  Da ket hon   

        Viec-lam  Tin-dung  
0  Dang lam viec    750.00  
1    That nghiep    680.00  
2  Dang lam viec    800.00  
3      Sinh vien    600.00  
4  Dang lam viec    720.00

## 7.2.2. Chuyển đổi dữ liệu phân loại sang dữ liệu số



Chuyển đổi dữ liệu sang dạng số là một phần quan trọng trong quá trình tinh chỉnh dữ liệu và xử lýdữ liệu trong khoa học dữ liệu và học máy. Điều này là cần thiết bởi vì nhiều mô hình học máy yêucầu dữ liệu đầu vào dưới dạng số thay vì dữ liệu dạng phân loại. Bên cạnh đó, khi dữ liệu chuyển đổidữ liệu sang dạng số, chúng ta có thể cải thiện hiệu suất của mô hình vì nó có khả năng xử lý cácbiến số tốt hơn, hiểu các mối quan hệ số hơn và thực hiện các phép tính toán dễ dàng hơn.

Dưới đây là một số các kỹ thuật phổ biến được sử dụng để chuyển đổi dữ liệu từ dạng phân loạisang dạng số
- Label Encoding: Mã hóa biến phân loại thành các giá trị số tương ứng. Ví dụ, biến "Goi-tinh"có thể được mã hóa thành 0 và 1, "Nam"và "Nu"tương ứng
- Ordinal Encoding: Mã hóa biến phân loại dạng có thứ tự. Ví dụ: Ví dụ, các giá trị "Daihoc,""Cao Dang"và "Trung hoc"có thể được thay thế thành các giá trị số phản ánh thứ tự này("Dai hoc": 2, "Cao dang": 1, "Trung hoc": 0)
- One-Hot Encoding: Tạo các cột giá trị nhị phân (0 hoặc 1) cho từng giá trị duy nhất trong biếnhạng mục. Điều này giúp mô hình xem xét từng giá trị riêng biệt. Ví dụ, biến "Viec-lam"có thể được chia thành các cột như "Dang lam viec", "That nghiep"và "Sinh vien"

### **Label Encoding**

Ví dụ, chúng ta cần thực hiện mã hóa giới tính sang dạng số 0,1 bằng cách thay các giá trị "Nu"thành"0"và các giá trị "Nam"thành

In [56]:
import pandas as pd
df = pd.read_csv("clean_khach_hang.csv")

df['Gioi-tinh'] = df['Gioi-tinh'].replace({"Nam":0, "Nu":1})

df.head()

/tmp/ipython-input-1961791910.py:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Gioi-tinh'] = df['Gioi-tinh'].replace({"Nam":0, "Nu":1})


Ma        Ho-ten  Tuoi  Gioi-tinh   Thu-nhap   Trinh-do    Hon-nhan  \
0   1  Nguyen Van A    30          0 5000000.00    Dai hoc  Da ket hon   
1   2    Tran Thi B    28          1 4500000.00   Cao dang  Da ket hon   
2   3     Le Minh C    35          0 6000000.00    Dai hoc  Da ket hon   
3   4    Phan Thi D    22          1 3500000.00   Cao dang    Doc than   
4   5   Hoang Van E    40          0 8000000.00  Trung hoc  Da ket hon   

        Viec-lam  Tin-dung  
0  Dang lam viec    750.00  
1    That nghiep    680.00  
2  Dang lam viec    800.00  
3      Sinh vien    600.00  
4  Dang lam viec    720.00

### **Ordinal Encoding**

Ví dụ, chúng ta cần thực hiện mã hóa cột "Trinh-do"sang dữ liệu số với các giá trị tương ứng là"Trung hoc": 0, "Cao dang": 1, "Dai hoc": 2. Trước tiên, chúng ta cần xây dựng bảng ánh xạ giátrị phân loại vào số tương ứng (ordinal_mapping). Sau đó, thực hiện quá trình ánh xạ dữ liệu bằngphương thức map() trong pandas

In [57]:
ordinal_mapping = {"Trung hoc":0, "Cao dang":1, "Dai hoc": 2}

df['Trinh-do'] = df['Trinh-do'].map(ordinal_mapping)

df.head()

Ma        Ho-ten  Tuoi  Gioi-tinh   Thu-nhap  Trinh-do    Hon-nhan  \
0   1  Nguyen Van A    30          0 5000000.00         2  Da ket hon   
1   2    Tran Thi B    28          1 4500000.00         1  Da ket hon   
2   3     Le Minh C    35          0 6000000.00         2  Da ket hon   
3   4    Phan Thi D    22          1 3500000.00         1    Doc than   
4   5   Hoang Van E    40          0 8000000.00         0  Da ket hon   

        Viec-lam  Tin-dung  
0  Dang lam viec    750.00  
1    That nghiep    680.00  
2  Dang lam viec    800.00  
3      Sinh vien    600.00  
4  Dang lam viec    720.00

### **One-Hot Encoding**

Ví dụ, chúng ta thực hiện mã hóa cột "Viec-lam"có các giá trị "Đang làm việc", "Thất nghiệp", và"Sinh viên"sử dụng phương pháp One-Hot Encoding thông qua phương thức get_dummies() trongpandas để tạo ra các cột giả tương ứng các nhóm giá trị của biến phân loại

In [58]:
df = pd.get_dummies(df, columns=['Viec-lam'])

df.head()

Ma        Ho-ten  Tuoi  Gioi-tinh   Thu-nhap  Trinh-do    Hon-nhan  \
0   1  Nguyen Van A    30          0 5000000.00         2  Da ket hon   
1   2    Tran Thi B    28          1 4500000.00         1  Da ket hon   
2   3     Le Minh C    35          0 6000000.00         2  Da ket hon   
3   4    Phan Thi D    22          1 3500000.00         1    Doc than   
4   5   Hoang Van E    40          0 8000000.00         0  Da ket hon   

   Tin-dung  Viec-lam_Dang lam viec  Viec-lam_Sinh vien  Viec-lam_That nghiep  
0    750.00                    True               False                 False  
1    680.00                   False               False                  True  
2    800.00                    True               False                 False  
3    600.00                   False                True                 False  
4    720.00                    True               False                 False

## 7.2.3. Chuyển đổi dữ liệu số về dạng phân loại

Chuyển dữ liệu số sang dạng phân loại (categorical data) là một phần quan trọng của qúa trình tinhchỉnh dữ liệu cho phân tích dữ liệu và một số các mô hình máy học. Quá trình này cũng được gọi làphân giỏ dữ liệu (data bining).

Một số trường hợp cần chuyển đổi dữ liệu số sang dạng phân loại:

Mục tiêu của phân tích: Đôi khi, mục tiêu của phân tích yêu cầu dự đoán phân lớp một kháchhàng. Ví dụ, dự đoán rủi ro tính dụng của một khách hàng ("cao", "Trung bình", "Thấp").

•Trực quan hóa dữ liệu: Dữ liệu dạng phân loại thường dễ trực quan hóa hơn. Bạn có thể sửdụng biểu đồ cột, biểu đồ tròn, hoặc biểu đồ thanh để hiển thị phân phối của các nhóm. Ví dụ,gom nhóm các khách hàng theo thu nhập ("Cao", "Trung bình", "Thấp").

•Giảm ảnh hưởng của giá trị cụ thể: Khi dữ liệu số được chuyển sang thành dạng phân loại ởcác nhóm, có thể giảm bớt ảnh hưởng của giá trị cụ thể (outliers) và giúp tạo ra các biểu đồ dễđọc hơn.

Ví dụ, chúng ta muốn phân giá trị của cột "Thu-nhap"thành các nhóm giá trị "Thấp", "Trungbình", và "Cao". Ở đây, chúng ta có 3 nhóm ("Thấp", "Trung bình", "Cao") (hoặc có thể gọi là 3giỏ), do đó cần 4 mốc gía trị để thực hiện quá trình phân chia. Ví dụ chúng ta chọn giá trị thấp nhấtcho cột thu nhập là 0, lớn nhất là vô cùng (float("inf")), và hai mốc nữa 5000000 và 7000000

Công thức phân chia như sau:
> Thấp: thu nhâp từ 0 đến 5000000

> Trung bình: thu nhâp từ 5000000 đến 7000000

> Cao: thu nhập trên 7000000

Trước tiên, chúng ta tạo ra các giỏ (bins) bằng cách định nghĩa khoảng giá trị cho từng nhóm.Tiếp theo, thiết lập nhãn (labels) cho từng nhóm dữ liệu. Cuối cùng, sử dụng phương thức cut() trong pandas để thực hiện phân nhóm cột "Thu-nhap

In [59]:
import  pandas  as pd

df = pd.read_csv("clean_khach_hang.csv")

bins = [0, 5000000, 7000000, float("inf")]

labels = ["Thap", "Trung binh", "Cao"]

df["Thu-nhap"] = pd.cut(df['Thu-nhap'], bins = bins,
                        labels = labels, right = False)


In [60]:
df.head()

Ma        Ho-ten  Tuoi Gioi-tinh    Thu-nhap   Trinh-do    Hon-nhan  \
0   1  Nguyen Van A    30       Nam  Trung binh    Dai hoc  Da ket hon   
1   2    Tran Thi B    28        Nu        Thap   Cao dang  Da ket hon   
2   3     Le Minh C    35       Nam  Trung binh    Dai hoc  Da ket hon   
3   4    Phan Thi D    22        Nu        Thap   Cao dang    Doc than   
4   5   Hoang Van E    40       Nam         Cao  Trung hoc  Da ket hon   

        Viec-lam  Tin-dung  
0  Dang lam viec    750.00  
1    That nghiep    680.00  
2  Dang lam viec    800.00  
3      Sinh vien    600.00  
4  Dang lam viec    720.00

# 7.3. TỔNG KẾT

Chương "Làm sạch và tinh chỉnh dữ liệu"đóng vai trò quan trọng trong quá trình xử lý dữ liệu đểđảm bảo chất lượng và độ tin cậy. Các kỹ thuật làm sạch dữ liệu bao gồm xử lý dữ liệu trùng lắp,thiếu, sai định dạng và giá trị. Kỹ thuật tinh chỉnh dữ liệu tập trung vào việc chuyển đổi giữa dữ liệusố và phân loại, tạo ra sự đồng nhất và thuận tiện cho quá trình phân tích.

Chương này cung cấp những bước cụ thể và công cụ để làm sạch và tinh chỉnh dữ liệu, giúpngười đọc hiểu rõ về các thách thức phổ biến khi làm việc với dữ liệu thô. Bằng cách này, nó tạo ramột nền tảng vững chắc cho chương tiếp theo về "Phân tích thăm dò,"nơi mà dữ liệu đã được chuẩnbị hoàn hảo để tìm hiểu và khám phá thông tin quan trọn

# 7.4  Bài tập



1.Thực hiện chuyển đổi các thuộc tính Giới tính, Trình độ, Hôn nhân và Tình trạng việc làmtrong tập dữ liệu "clean_khach_hang.csv"về thành dạng số.

2.Chuyển thuộc tính Tuổi thành dạng phân loại với 3 giá trị: trẻ (x < 25), trung niên (25 <= x<35) và lớn tuổi (x > 35).

3.Thực hiện chuyển đổi cột điểm Tín dụng về thành dạng biến phân loại với 2 giá trị: cao vàthấp